koku/koku/pg_partition.py

```diff
@@ -1066,3 +1070,281 @@ def drop_original_objects(self):
+        partition_start_sql = f"""
+SELECT DISTINCT
+       date_trunc('month', usage_start)::date as "partition_start",
+       date_trunc('month', (date_trunc('month', usage_start) + '1 month'::interval))::date as "partition_end"
+  FROM "{self.schema_name}"."{self.default_partition}"
+ ORDER
+    BY 1;
+"""
+        LOG.info(f'Getting partition bounds for data in "{self.schema_name}"."{self.default_partition}"')
+        cur = conn_execute(partition_start_sql)
+        res = [(r["partition_start"], r["partition_end"]) for r in fetchall(cur)]
+
+        if len(res) == 0:
+            LOG.info("Nothing to do")
+
+        return res
+
+    def repartition_default_data(self):
+        new_partitions = self._get_new_partitions_from_default()
+        if new_partitions:
+            mv_recs_sql = f"""
+WITH __mv_recs_{self.tx_id} as (
+DELETE
+  FROM "{self.schema_name}"."{self.default_partition}"
+ WHERE usage_start >= %s
+   AND usage_start < %s
+RETURNING *
+)
+INSERT INTO {{}}
+SELECT * FROM __mv_recs_{self.tx_id} ;
+"""
+
+            for bounds in new_partitions:
+                p_from, p_to = bounds
+                self.partition_name = f"{self.partitioned_table}_{p_from.strftime('%Y_%m')}"
+                full_partition_name = f'"{self.schema_name}"."{self.partition_name}"'
+
+                # A little jiggery-pokery here to increase efficiency.
+                # The partition is named properly, but it set as a partition for 100 years in the future
+                # This will allow us to create the partition with all of the structure that is required.
+                # Before the data move, we will detach the partition, then move the data,
+                # then reattach with the proper bounds.
+                self.partition_parameters = {
+                    "from": p_from.replace(year=p_from.year + 100),
+                    "to": p_to.replace(year=p_to.year + 100),
+                }
+
+                LOG.info(
+                    f"Re-partitioning {self.schema_name}.{self.default_partition} into {full_partition_name} "
+                    + f"for {self.partition_key} values from {p_from} to {p_to}"
+                )
+                with transaction.atomic():
+                    # Create new partition
+                    self._create_partition()
+                    # Detach it so we can move the data
+                    self._detach_partition()
+
+                    # Move data into new partition
+                    conn_execute(mv_recs_sql.format(full_partition_name), (p_from, p_to), _conn=self.conn)
+
+                    # Re-attach partition with actual bounds
+                    self.partition_parameters["from"] = p_from
+                    self.partition_parameters["to"] = p_to
+                    self._attach_partition()
+
+
+# This is a crawler interface to the PartitionDefaultData class
+def repartition_default_data(schema_name=None, partitioned_table_name=None):
```

Comment:
This is a crawler that will gather default partition info from across all of our tenants for processing all at once. It can also be constrained to a single schema, a single partitioned table or both.